frame/epoch-vrf/src/lib.rs

```diff
@@ -438,72 +445,38 @@ impl<T: Trait> Module<T> {
 			return;
 		}
 
-		let maybe_pre_digest: Option<PreDigest> = <frame_system::Module<T>>::digest()
-			.logs
-			.iter()
-			.filter_map(|s| s.as_pre_runtime())
-			.filter_map(|(id, mut data)| if id == BABE_ENGINE_ID {
-				PreDigest::decode(&mut data).ok()
-			} else {
-				None
-			})
-			.next();
+		let maybe_pre_digest: Option<T::PreDigest> = T::find_pre_digest();
 
 		let maybe_randomness: Option<schnorrkel::Randomness> = maybe_pre_digest.and_then(|digest| {
 			// on the first non-zero block (i.e. block #1)
 			// this is where the first epoch (epoch #0) actually starts.
 			// we need to adjust internal storage accordingly.
-			if GenesisSlot::get() == 0 {
-				GenesisSlot::put(digest.slot_number());
-				debug_assert_ne!(GenesisSlot::get(), 0);
+			if <GenesisSlot<T>>::get() == Default::default() {
+				<GenesisSlot<T>>::put(digest.slot_number());
+				#[cfg(feature = "std")] {
+					debug_assert_ne!(<GenesisSlot<T>>::get(), Default::default());
+				}
 
 				// deposit a log because this is the first block in epoch #0
 				// we use the same values as genesis because we haven't collected any
 				// randomness yet.
-				let next = NextEpochDescriptor {
-					authorities: Self::authorities(),
-					randomness: Self::randomness(),
-				};
-
-				Self::deposit_consensus(ConsensusLog::NextEpochData(next))
+				T::deposit_next_epoch_descriptor(Self::authorities(), Self::randomness());
 			}
 
 			// the slot number of the current block being initialized
 			let current_slot = digest.slot_number();
 
 			// how many slots were skipped between current and last block
-			let lateness = current_slot.saturating_sub(CurrentSlot::get() + 1);
-			let lateness = T::BlockNumber::from(lateness as u32);
+			let lateness = current_slot.saturating_sub(<CurrentSlot<T>>::get() + One::one());
 
-			Lateness::<T>::put(lateness);
-			CurrentSlot::put(current_slot);
+			Lateness::<T>::put(UniqueSaturatedInto::<T::BlockNumber>::unique_saturated_into(lateness));
+			CurrentSlot::<T>::put(current_slot);
 
-			if let PreDigest::Primary(primary) = digest {
+			if let Some(randomness) = T::make_randomness(&digest) {
 				// place the VRF output into the `Initialized` storage item
 				// and it'll be put onto the under-construction randomness
 				// later, once we've decided which epoch this block is in.
```

Comment:
Can someone explain the issue with knowing in which epoch the block comes?  It'd be easier on the storage if `UnderConstruction = Hash(UnderConstruction, BlockVrfOutput)` but you cannot do that if you donnot what epoch a block lies in.  We should always know what epoch sassafras blocks lie in though, yes?

Reply:
It's just that at this point we are still running the block initialization logic and haven't processed any epoch change. It might be the case that this block starts a new epoch, in which case the randomness should be pooled under the new epoch (i.e. empty pool) rather than adding it to the existing epoch's. This will all be done as part of block processing so once we exit from the runtime the randomness will be stored in its appropriate place (you're right that we can always tell what epoch a given block belongs to).

Reply:
In sassafras, we authorized block production by building this VRF output list for the whole epoch, so we'll need to know our epoch so that we can check the right list.  I suspect this might just be a place where babe and sassafras diverge more than expected.